### PR TITLE
Prevent reallocation of aligned memory

### PR DESCRIPTION
**Pull request by mlxd**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      [`tests`](../tests) directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `.github/CHANGELOG.md` file, summarizing the
      change, and including a link back to the PR.

- [x] Ensure that code is properly formatted by running `make format`. 

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:** To ensure the memory allocated by Lightning is always aligned, and usable by the dispatcher AVX kernels, we must avoid any potential numpy-layer reallocations. This PR extends the aligned allocator functions bound through Pybind11 to allow data to be initialized and set for use. This avoids us needing to perform potential numpy casts, which could potentially reallocate the data. As a result, this removes overheads observed with lightning.qubit workloads depending on the need to handle multiple resets/allocations.

**Description of the Change:** As above.

**Benefits:** Removes unneeded overheads from lightning.qubit, and redundant imports from LG/LK

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane_lightning/core/src/utils/Memory.hpp [hunk 1]
```diff
@@ -35,23 +36,28 @@ namespace Pennylane::Util {
...
-    return _aligned_malloc(bytes, alignment);
+    p = _aligned_malloc(bytes, alignment);
 #else
-    return std::aligned_alloc(alignment, bytes);
+    p = std::aligned_alloc(alignment, bytes);
 #endif
+    if (zero_init) {
+        std::memset(p, 0, bytes);
+    }
+    return p;
```
Comment (maliasadi) [inline pennylane_lightning/core/src/utils/Memory.hpp:60]: Nice addition, but I cannot find any actual use case of  `zero_init = True` in the codebase? Is there a followup PR?

Comment (mlxd) [reply pennylane_lightning/core/src/utils/Memory.hpp:60]: And this is why code-reviews are so important: The following chunk is missing:
```
        def _create_basis_state(self, index):
            """Return a computational basis state over all wires.
            Args:
                index (int): integer representing the computational basis state
            Returns:
                array[complex]: complex array of shape ``[2]*self.num_wires``
                representing the statevector of the basis state
            Note: This function does not support broadcasted inputs yet.
            """
            state = allocate_aligned_array(2**self.num_wires, np.dtype(self.C_DTYPE), True)
            state[index] = 1
            return self._reshape(state, [2] * self.num_wires)
```

I'll add that in.

Comment (mlxd) [reply pennylane_lightning/core/src/utils/Memory.hpp:60]: And done.

File: pennylane_lightning/core/src/bindings/Bindings.hpp [hunk 2]
```diff
@@ -172,20 +173,23 @@ auto alignedNumpyArray(CPUMemoryModel memory_model, size_t size) -> py::array {
  * @param size Size of the array to create
  * @param dt Pybind11's datatype object
  */
-auto allocateAlignedArray(size_t size, const py::dtype &dt) -> py::array {
+auto allocateAlignedArray(size_t size, const py::dtype &dt,
```
Comment (maliasadi) [inline pennylane_lightning/core/src/bindings/Bindings.hpp:176]: What do you think of adding `std::memset` to this method to remove `zeroInit` argument with some default value from the rest of C++ codebase? I assume `zeroInit` is mainly useful in this method and removing it from other places would reduce the number of code changes while keeping the signature of those methods simple.

File: pennylane_lightning/core/_version.py [hunk 3]
```diff
@@ -16,4 +16,4 @@
    Version number (major.minor.patch[-label])
```
Comment (maliasadi) [inline pennylane_lightning/core/_version.py:16]: Would you update the changelog :)

File: pennylane_lightning/core/src/bindings/Bindings.hpp (hunk 2 above)
Comment (mlxd) [reply pennylane_lightning/core/src/bindings/Bindings.hpp:176]: Happy to in a follow up PR.

File: pennylane_lightning/core/_version.py (hunk 3 above)
Comment (mlxd) [reply pennylane_lightning/core/_version.py:16]: Will do once approved

**Review by maliasadi — APPROVED** [review]
Thanks @mlxd! Looks good to go :t-rex:

File: pennylane_lightning/core/src/bindings/Bindings.hpp (hunk 2 above)
Comment (mlxd) [reply pennylane_lightning/core/src/bindings/Bindings.hpp:176]: TODO added.

File: requirements-dev.txt [hunk 4]
```diff
@@ -1,4 +1,4 @@
-pip==23.0
+pip~=22.0
```
Comment (multiphaseCFD) [inline requirements-dev.txt:1]: Want to know the reason why we use ~=22.0 here instead of 23.0

**Review by multiphaseCFD — APPROVED** [review]
Looks good to me! Thanks @mlxd !

File: requirements-dev.txt (hunk 4 above)
Comment (mlxd) [reply requirements-dev.txt:1]: The latest version of the 23.x branch causes issues with pip-installed binaries being available and usable on path. I downgraded the previous major release for now, until we can identify the best path around this.

**Pull request merged**